### PR TITLE
Load module autoloaders and service configurations before the core Symfony compiler passes are executed

### PR DESCRIPTION
**Pull request by dr-woland**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This PR modifies the order in which the service containers are compiled in a way that should allow module developers to use more of the configuration options provided by the Symfony framework when defining services. See below for details.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | -
| Related PRs       | -
| How to test?      | Example modules that try to demonstrate the mentioned topics are attached to this PR - details below.
| Possible impacts? | The changes should not impact the previously working service configurations, but it probably would not hurt to check a few example modules relying on Symfony services to verify that they are working as intended. Hopefully, the worst that happens is some left and forgotten `_instanceof` configurations suddenly starting to work and apply options that may not be desired anymore... ;)

### Description

Symfony 3.3 introduced several features that allow to simplify service configuration:
- [applying configuration based on the extended classes or implemented interfaces](https://symfony.com/blog/new-in-symfony-3-3-simpler-service-configuration#interface-based-service-configuration)
- [autoconfiguration](https://symfony.com/blog/new-in-symfony-3-3-simpler-service-configuration#interface-based-service-configuration)
- [an option to skip the `class` attribute](https://symfony.com/blog/new-in-symfony-3-3-optional-class-for-named-services)
- [automatically registering classes found in the specified directories as services](https://symfony.com/blog/new-in-symfony-3-3-psr-4-based-service-discovery)

Unfortunately, it is not currently possible to use any of the above-mentioned when configuring services for the PrestaShop modules. The reason for that is because the `PrestaShopBundle\DependencyInjection\Compiler\LoadServicesFromModulesPass`, which is responsible for loading the module service definitions, is executed after `Symfony\Component\DependencyInjection\Compiler\ResolveClassPass` and `Symfony\Component\DependencyInjection\Compiler\ResolveInstanceofConditionalsPass` that are responsible for setting the `class` attribute in the service definition if missing and applying configuration depending on `instanceof` conditionals. This PR addresses this issue by increasing the priority of the PrestaShop's compiler pass from 1 to 200 (it has to be greater than 100, which is the priority for both of the core Symfony passes).

Furthermore, because classes defined in module files might not exist during the service container compilation in the legacy environments (module autoloaders are not automatically included until after the container is compiled - in contrast to the BO, where it happens before container initialization in the `AppKernel`), some additional features that do work on the BO (e.g. autowiring) are not available on the FO (technically, it is possible to work around the issue by registering one of the hooks executed before controller initialization and including your classes/autoloader in the file declaring the module class...) and in the WebService context (not without overrides, which is an even uglier - and more unreliable at that - hack). This problem is fixed by adding the `loadModulesAutoloader` method call in `PrestaShop\PrestaShop\Adapter\ContainerBuilder` before calling `compile` on the core Symfony container builder (autloaders are still included separately if the cache is fresh). 

### How to test?

I have prepared 2 simple modules to help test the changes:

[classresolutiontest.zip](https://github.com/PrestaShop/PrestaShop/files/10240586/classresolutiontest.zip)
The module contains a few classes:
classresolutiontest
and a simple service configuration:
```yaml
# config/common.yml
services:
  _defaults:
    public: true
  TestModule\ClassResolution\:
    resource: ../src
    exclude:
      - ../src/ExcludedDirectory
      - ../src/**/index.php
  TestModule\ClassResolution\ExcludedDirectory\SkippedClassKeyInServiceDefinition: ~
```
3 messages saying whether or not a given class is present in the container as a service (2 out of 3 should be) are printed on the module's configuration page and on the home page. Without the changes from this PR, an exception with a somewhat misleading message is thrown during container compilation:
no_class_attribute

---

[instanceofconditionalstest.zip](https://github.com/PrestaShop/PrestaShop/files/10240587/instanceofconditionalstest.zip)
The module contains a controller, a collection class and an element class implementing an interface:
instanceofconditionalstest
Service configuration contains 2 definitions of collection services each autowiring all of the specifically tagged services (the collection constructor argument is [bound by name](https://symfony.com/doc/current/service_container.html#binding-arguments-by-name-or-type)):
```yaml
# config/common.yml
services:
  _defaults:
    autowire: true
    bind:
      $elements: !tagged test_module.instance_of.instance_of_tagged
  TestModule\InstanceofConditionals\Collection\Collection:
    class: TestModule\InstanceofConditionals\Collection\Collection
    public: true
  test_module.instance_of.manually_tagged_collection:
    class: TestModule\InstanceofConditionals\Collection\Collection
    public: true
    bind:
      $elements: !tagged test_module.instance_of.manually_tagged
```
as well as an element service that should have 2 tags (1 configured directly, 1 based on an `instanceof` conditional):
```yaml
  _instanceof:
    TestModule\InstanceofConditionals\Collection\ElementInterface:
      tags: [ test_module.instance_of.instance_of_tagged ]
  TestModule\InstanceofConditionals\Collection\Element:
    class: TestModule\InstanceofConditionals\Collection\Element
    tags: [ test_module.instance_of.manually_tagged ]
```
The controller service definition is both autowired and autoconfigured.
The module configuration page displays links to 2 controller actions:
* The first one displays the count of elements in each of the collection services. With the changes from this PR, both collections should contain a single element. Without them one is empty.
* The second takes an optional parameter and displays a different message depending if the argument was provided or not. It should be autowired thanks to the `controller.service_arguments` tag (because the controller class extends the Symfony's `AbstractController` it should come from the autoconfiguration), but without the changes from this PR it is not.

Messages similar to the ones from the first controller action are displayed on the home page. But if the `moduleRoutes` hook is not enabled for the module (its sole purpose is including the autoloader), an exception is thrown when trying to compile the container:
unused_binding

---

You could also experiment with other options that you might find in [the Symfony documentation](https://symfony.com/doc/4.4/service_container.html#creating-configuring-services-in-the-container).

### CODE REVIEW

File: src/Adapter/ContainerBuilder.php [hunk 1]
```diff
@@ -134,14 +134,14 @@ public function buildContainer($containerName)
         $container = $this->loadDumpedContainer();
         if (null === $container) {
             $container = $this->compileContainer();
+        } else {
+            $this->loadModulesAutoloader($container);
```
Comment (jolelievre) [inline src/Adapter/ContainerBuilder.php:139]: That's the only part I'm not sure to understand, why did you need to load this before we set `shop` end `employee` definitions?

File: src/Adapter/ContainerBuilder.php [hunk 2]
```diff
@@ -170,7 +170,7 @@ private function compileContainer()
         //If the container builder is modified the container logically should be rebuilt
         $container->addResource(new FileResource(__FILE__));
 
-        $container->addCompilerPass(new LoadServicesFromModulesPass($this->containerName), PassConfig::TYPE_BEFORE_OPTIMIZATION, 1);
+        $container->addCompilerPass(new LoadServicesFromModulesPass($this->containerName), PassConfig::TYPE_BEFORE_OPTIMIZATION, 200);
```
Comment (jolelievre) [inline src/Adapter/ContainerBuilder.php]: Can you please add a comment here explaining why this specific priority is used, that it needs to have a higher priority than `ResolveClassPass` and `ResolveInstanceofConditionalsPass` (too bad this priority is hard-coded in the framework a const would have been nice to use as a reference)

File: src/PrestaShopBundle/PrestaShopBundle.php [hunk 3]
```diff
@@ -60,8 +60,8 @@ public function build(ContainerBuilder $container)
     {
         $container->addCompilerPass(new DynamicRolePass());
         $container->addCompilerPass(new PopulateTranslationProvidersPass());
-        $container->addCompilerPass(new LoadServicesFromModulesPass(), PassConfig::TYPE_BEFORE_OPTIMIZATION, 1);
-        $container->addCompilerPass(new LoadServicesFromModulesPass('admin'), PassConfig::TYPE_BEFORE_OPTIMIZATION, 1);
+        $container->addCompilerPass(new LoadServicesFromModulesPass(), PassConfig::TYPE_BEFORE_OPTIMIZATION, 200);
```
Comment (jolelievre) [inline src/PrestaShopBundle/PrestaShopBundle.php]: Same comment here to explain the priority, it's just so that a future developer understands why this specific value is used and doesn't change it

**Review by jolelievre — COMMENTED** [review]
Thanks @dr-woland 

Amazing contribution, and so many details in the description almost nothing to ask it's very clear! I had just a couple suggestion for additional comments but that's all

**Review by 0x346e3730 — DISMISSED** [review]
Amazing PR, I would suggest referring this PR in the comments to explain why the 200 priority because the PR gives a lot of context and explains everything really nice.

File: src/PrestaShopBundle/PrestaShopBundle.php (hunk 3 above)
Comment (matthieu-rolland) [reply src/PrestaShopBundle/PrestaShopBundle.php]: we could even add the # of the PR so that future devs can see the full explanation if needed

File: src/Adapter/ContainerBuilder.php (hunk 1 above)
Comment (dr-woland) [reply src/Adapter/ContainerBuilder.php:139]: The `shop` and `employee` services being set in the container or not should make no difference to the autoloaders. Previously, the `loadModulesAutoloader` method was called after injecting these instances into the container regardless of whether the container was loaded from the cache or recompiled. Now, it is only needed if the cache was fresh, because otherwise the autoloaders are included in the `compileContainer` method (and I wanted to avoid unnecessarily calling `loadModulesAutoloader` twice). We could introduce an additional variable storing the result of the `null === $container` conditional and add another `if` statement after injecting the synthetic services, but I do not think it is necessary.

File: src/Adapter/ContainerBuilder.php (hunk 2 above)
Comment (dr-woland) [reply src/Adapter/ContainerBuilder.php]: Thanks for the suggestion. The CS fixer did not like the phpDoc comment above the method call (because of the `phpdoc_to_comment` rule), so I added a new constant in the `PrestaShopBundle` class, added a comment there and replaced the hard-coded 200 priorities with the constant usage.

File: src/PrestaShopBundle/PrestaShopBundle.php (hunk 3 above)
Comment (dr-woland) [reply src/PrestaShopBundle/PrestaShopBundle.php]: I added a link to this PR in the doc comment.
https://github.com/PrestaShop/PrestaShop/pull/30588#discussion_r1050611213

**Review by kpodemski — APPROVED** [review]
I think there's a slight chance that changes may impact the third-party modules, but we'll perform QA by dev anyway.

Thanks @dr-woland

**Review by zuk3975 — APPROVED** [review]

**Comment by kpodemski** [comment]
As PR is accepted, I will test it next week to see if there's no regression with some open-source modules, and I'll test new features.

**Comment by kpodemski** [comment]
Hello @dr-woland 

Ok! I've tested it with the modules you provided, example modules from the project and some company modules, tinkered with the services configuration, and run automated tests. All seems to be OK! 

Thank you very much for your PR. I hope it's one of many! Well done 👏🏻

**Comment by thomasnares** [comment]
HI @dr-woland, i added your examples to the documentation, would you review/suggest me changes if needed on this PR ? 
https://github.com/PrestaShop/docs/pull/1623

Thanks again for this great contribution!
